chain/find.go

```diff
@@ -222,44 +222,42 @@ func (g *GapIndexer) findTaskEpochGaps(ctx context.Context) (visor.GapReportList
 		ctx,
 		&result,
 		`
-select height, task
-	from (
-
-	-- incomplete_heights_and_their_completed_tasks: for every incomplete
-	-- heightin incomplete_task_counts we will take all processing_reports
-	-- with the same height and return height,tasks which are 'OK'
-	-- which provides all incomplete heights and their completed tasks
-	select distinct vpr.task, vpr.height
-	from visor_processing_reports vpr
-	right join(
-
-		-- incomplete_task_counts: count of all tasks by height,status
-		-- and filter for status = 'OK' and all task counts which
-		-- don't equal len(AllTasks)
-		select incomplete_task_counts.height, status, count(incomplete_task_counts.height) cheight
-		from (
-
-			-- all unique height,tasks,status within range of interest
-			select task, height, status
-			from visor_processing_reports
-			where height >= ?2 and height <= ?3
-			group by height, task, status
-			order by height desc, task
-
-		) incomplete_task_counts
-		where status = ?4
-		group by incomplete_task_counts.height, incomplete_task_counts.status
-		having count(incomplete_task_counts.height) != ?0
-
-	) incomplete_heights
+with
+
+-- ranked_tasks: filters by height and prepares ranking status of tasks by int
+-- only considers heights with incomplete task count
+ranked_tasks as (
+	select height, task, status,
+	case status -- adding a col of int for statuses to make choosing tasks that have OK and non-OK status easier.
+		when 'OK' then 0
+		when 'INFO' then 1
+		when 'SKIP' then 2
+		when 'ERROR' then 3
+		else null
+	end as status_rank
+	from visor_processing_reports
+	where height >= ?2 and height <= ?3
+	group by height, task, status
+	having count(*) != ?0
+),
+
+-- dense_ranked_tasks: where the actual "ranking" happens
+-- rank by status_rank
+-- rank 1 means the lowest value status rank found of the tuple (height, task)
+dense_ranked_tasks as (
+	select *,
+	dense_rank() over (partition by height, task order by status_rank) as ranked_status
```

Comment:
A sample of  how the dense ranking happens:

```
height,task,status,ranked_status
890000,actorstatesinit,OK,1
890000,actorstatesinit,SKIP,2
890000,actorstatesmarket,OK,1
890000,actorstatesmarket,SKIP,2
890000,actorstatesminer,ERROR,2
890000,actorstatesminer,OK,1
```

By ranking on `status_rank` with partitions over `(height, task)`, we are ensuring that we are selecting in the set the lowest ranked task and any tasks that have had an `OK` (or whatever given) status.

Reply:
> we are ensuring that we are selecting in the set the _lowest_ ranked task

Where is this assertion happening? I only see a condition for `rt.ranked_status = 1` (which doesn't even need to be an int since matching on strings work just as well... and it's not clear if the transform+equivalency check on an int is going to be cheaper than just a text equivalency check on text.)

Reply:
It's on this [line](https://github.com/filecoin-project/lily/pull/884/files#diff-d49953a87209cc88edee91a3c0173f069297d6c4835cd1e17340094e4fbfd94aR258), but upon taking a look at it again, you're right. A simple text check would work just fine. I initially worked on this query with the assumption that we would be finding tasks do not have a single `OK` on it, hence the ranking.

Reply:
Yeah, I think using a weighted filter (via dense_rank) might be a little overkill here. If the absence of the 'OK' is the case you're attempting to solve for here, it would be more obvious as a LEFT JOIN selecting for NULL.

Here's my whack at this some time back which we decided we weren't ready for. Maybe there are some patterns there that might help you. (https://github.com/filecoin-project/lily/blob/48437e66bcf1c7b5f990de3f69d633e38b1353bf/chain/find.go#L88-L193)